### PR TITLE
Features/win32 allow cancel app shutdown

### PR DESCRIPTION
**Pull request by danwalmsley**
## What does the pull request do?

Adds support for WM_QUERYENDSESSIOn on win32 so that avalonia apps can prevent system shutdown.

## What is the current behavior?

Only supported on OSX,.

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

### CODE REVIEW

File: samples/Sandbox/App.axaml.cs [hunk 1]
```diff
@@ -16,7 +16,14 @@ public override void OnFrameworkInitializationCompleted()
             if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktopLifetime)
             {
                 desktopLifetime.MainWindow = new MainWindow();
+
+                desktopLifetime.ShutdownRequested += DesktopLifetime_ShutdownRequested;
             }
         }
+
+        private void DesktopLifetime_ShutdownRequested(object sender, System.ComponentModel.CancelEventArgs e)
```
Comment (grokys) [inline samples/Sandbox/App.axaml.cs:24]: This should be removed before merging.

**Review by grokys — COMMENTED** [review]
I tried this with `taskkill` to send a termination message to the application, and the `ShutdownRequested` handler didn't get called.

`WM_QUERYENDSESSION` is only sent when the user tries to shutdown the OS itself, which is different to the signal we're handling on OSX, which is for an _application_ shutdown.

If we're going to handle both cases in `ShutdownRequested` then I think we need an enum on the event args so an app author can tell what the event represents, because returning `False` in response to `WM_QUERYENDSESSION` has specific implications (i.e. your app is likely to be forcefully closed if you don't do anything for 5 seconds after returning `False`)

**Review by kekekeks — APPROVED** [review]

**Comment by danwalmsley** [comment]
> `WM_QUERYENDSESSION` is only sent when the user tries to shutdown the OS itself, which is different to the signal we're handling on OSX, which is for an _application_ shutdown.

The OSX one gets called for both, and is supposed to be used for preventing shutdown also.

You can even resume shutting down by returning `NSTerminateLater` from that function and calling another api later...

File: samples/Sandbox/App.axaml.cs (hunk 1 above)
Comment (danwalmsley) [reply samples/Sandbox/App.axaml.cs]: removed... just wanted to show how it would be used.

**Comment by danwalmsley** [comment]
@grokys iv updated the documentation for the event, to clarify its role on OSX and Win32..

I beleive this should be good now.

**Comment by Gillibald** [comment]
So there are multiple ways to signal an application / window to terminate gracefully: https://devblogs.microsoft.com/oldnewthing/20190411-00/?p=102411

**Comment by danwalmsley** [comment]
discussed with @grokys offline:

- At this stage a reason for shutdown isnt required and difficult to abstract on multiple platforms.
- Added a special CancelEventArgs class that we can add to later without breaking api compat.

**Review by grokys — APPROVED** [review]

**Pull request merged**